zenith_utils/src/chunked_buffer.rs

```diff
@@ -0,0 +1,173 @@
+use std::cmp::min;
+use std::io::{Read, Write};
+
+const CHUNK_SIZE: usize = 1024;
+
+///
+/// ChunkedBuffer is an expandable byte buffer. You can append to the end of it, and you
+/// can read at any byte position. The advantage over a plain Vec<u8> is that the
+/// buffer consists of smaller chunks, so that a big buffer doesn't require one huge
+/// allocation, and expanding doesn't require copying all the data.
+///
+#[derive(Debug, Default)]
+pub struct ChunkedBuffer {
+    chunks: Vec<Box<[u8; CHUNK_SIZE]>>,
+    len: usize,
+}
+
+impl ChunkedBuffer {
+    /// Return current length of the buffer, in bytes
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Create a "cursor" for reading from the buffer.
+    pub fn reader(&self, pos: u64) -> ChunkedBufferReader {
+        ChunkedBufferReader {
+            parent: self,
+            pos: pos as usize,
+        }
+    }
```

Comment:
I think you should do the same thing with writer, and give out a "ticket" to the written range